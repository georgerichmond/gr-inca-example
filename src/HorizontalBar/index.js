// @flow
import { Chart, Coord, Geom, Legend } from 'bizcharts';
import { DataView } from '@antv/data-set';
import React from 'react';
import * as R from 'ramda';

type Props = {
  data: [
    {
      name: string,
      value: number
    }
  ]
};

const formatPercent = val => Math.round(val * 100) + '%';
const chartProps = { padding: 'auto' };

const sortByNameCaseInsensitive = R.sortBy(
  R.compose(
    R.toLower,
    R.prop('name')
  )
);

export const HorizontalBar = ({ data }: Props) => {
  const dv = new DataView();
  dv.source(sortByNameCaseInsensitive(data)).transform({
    type: 'percent',
    field: 'value',
    dimension: 'name',
    as: 'percent'
  });
  const cols = {
    percent: {
      formatter: formatPercent
    }
  };
  const itemFormatter = name =>
    `${name.replace(
      /\w+/g,
      word => word.charAt(0).toUpperCase() + word.substr(1)
    )} (${formatPercent(
      R.pathOr({}, ['percent'], dv.rows.find(point => point.name === name))
    )})`;

  return (
    <Chart height={50} {...chartProps} forceFit data={dv} scale={cols}>
      <Legend
        offsetY={-10}
        textStyle={{ fill: 'rgba(255, 255, 255, 0.65)' }}
        itemFormatter={itemFormatter}
      />
      <Coord transpose />
      <Geom
        type="intervalStack"
        position="1*percent"
        color={['name']}
        size={10}
      />
    </Chart>
  );
};

export default HorizontalBar;
