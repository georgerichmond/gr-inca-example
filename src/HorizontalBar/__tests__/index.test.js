import React from 'react';
import { shallow } from 'enzyme';
import HorizontalBar from '../index';

describe('HorizontalBar', () => {
  describe('when given data', () => {
    test('it renders correctly', () => {
      const props = {
        data: [
          {
            name: 'RA',
            value: 10
          },
          {
            name: 'subscriber',
            value: 50
          },
          {
            name: 'anonymous',
            value: 40
          }
        ],
        name: 'Unique visitors',
        metric: 'uniqueVisitors',
        interval: '1h'
      };
      const tree = shallow(<HorizontalBar {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });
});
