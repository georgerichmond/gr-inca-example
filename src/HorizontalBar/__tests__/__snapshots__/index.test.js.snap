// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalBar when given data it renders correctly 1`] = `
<BChart
  data={
    View {
      "_events": Object {
        "change": Array [],
      },
      "_source": Object {
        "options": undefined,
        "source": Array [
          Object {
            "name": "anonymous",
            "value": 40,
          },
          Object {
            "name": "RA",
            "value": 10,
          },
          Object {
            "name": "subscriber",
            "value": 50,
          },
        ],
      },
      "dataSet": null,
      "dataType": "table",
      "isDataView": true,
      "isView": true,
      "loose": true,
      "origin": Array [
        Object {
          "name": "anonymous",
          "value": 40,
        },
        Object {
          "name": "RA",
          "value": 10,
        },
        Object {
          "name": "subscriber",
          "value": 50,
        },
      ],
      "rows": Array [
        Object {
          "name": "anonymous",
          "percent": 0.4,
          "value": 40,
        },
        Object {
          "name": "RA",
          "percent": 0.1,
          "value": 10,
        },
        Object {
          "name": "subscriber",
          "percent": 0.5,
          "value": 50,
        },
      ],
      "transforms": Array [
        Object {
          "as": "percent",
          "dimension": "name",
          "field": "value",
          "type": "percent",
        },
      ],
      "watchingStates": null,
    }
  }
  forceFit={true}
  height={50}
  padding="auto"
  scale={
    Object {
      "percent": Object {
        "formatter": [Function],
      },
    }
  }
>
  <Legend
    itemFormatter={[Function]}
    offsetY={-10}
    textStyle={
      Object {
        "fill": "rgba(255, 255, 255, 0.65)",
      }
    }
    visible={true}
  />
  <TypedComponent
    transpose={true}
  />
  <TypedComponent
    color={
      Array [
        "name",
      ]
    }
    position="1*percent"
    size={10}
    type="intervalStack"
  />
</BChart>
`;
